src/htm/algorithms/SDRClassifier.cpp

```diff
@@ -165,13 +169,12 @@ void Predictor::reset() {
 }
 
 
-Predictions Predictor::infer(const UInt recordNum, const SDR &pattern)
-{
-  updateHistory_( recordNum, pattern );
+Predictions Predictor::infer(const UInt recordNum, const SDR &pattern) const {
+  checkMonotonic_(recordNum);
```

Comment:
@ctrl-z-9000-times please have a look on the last 1,(2) commits, I intended to make `infer` const, as it imho should have been. For Classifier that was easy, for Predictor I had to remove the `updateHistory_` from infer (no tests visibly broken).

Reply:
Why does it matter if infer is constant?  I doubt it will have any performance impact, and I don't think this will prevent any programming mistakes.

Consider the following chain of events:
```
infer( t, SDR-A ) -> PDF
learn( t+1, SDR-B, Labels )
```
The method `updateHistory` stores the given SDR inside of the predictor.  Previously the call to `learn` would have associated `SDR-A` with `Labels` since that SDR was given to infer with.  Now that will not happen.

Also, these changes allow the timestamps to the infer method to go backwards.
```
infer( t + 2, ... )
infer( t, ... )
```

I'm not saying the new behavior is *wrong*, just that it changed.

Reply:
>  I doubt it will have any performance impact, and I don't think this will prevent any programming mistakes.

right, there won't be any performance gains, and the behavior has changed. I think it adds (expected) semantics, and better separates responsibilities of those 2 functions: 
- you know only what you `learn()` 
- and you can `infer() const` anytime without any worry of changing the state (comes with loosened requirement for monotonic timestamps)

Reply:
Ok, that makes sense +1